src/persist-client/src/examples/persistent_source_example.rs

```diff
@@ -391,6 +429,249 @@ mod impls {
+                current_version: Arc::new(tokio::sync::Mutex::new(0)),
+            }
+        }
+    }
+
+    #[async_trait]
+    impl<ST, C> Timestamper for ConsensusTimestamper<ST, C>
+    where
+        ST: timely::PartialOrder + Sync + Send + Debug + Copy + Hash,
+        C: IteratedConsensus<ST, u64, u64> + Sync + Send,
+    {
+        type SourceTimestamp = ST;
+        type AssignedTimestamp = u64;
+
+        async fn get_bindings(
+            &self,
+            lower: Antichain<Self::SourceTimestamp>,
+            upper: Antichain<Self::SourceTimestamp>,
+        ) -> TimestampBindings<Self::SourceTimestamp, Self::AssignedTimestamp> {
+            let mut result_bindings = Vec::new();
+
+            let mut cache = self.cache.lock().await;
+            let mut current_version = self.current_version.lock().await;
+
+            for upper_element in upper.iter() {
+                let cached = cache.entry(*upper_element);
+
+                match cached {
+                    Entry::Occupied(cached_binding) => {
+                        println!("Cached binding: {:?}", cached_binding.get());
+                        result_bindings.push(*cached_binding.get());
+                        continue;
+                    }
+                    Entry::Vacant(e) => {
+                        // Well, this is awkward: there's no binding yet so we have to try and mint
+                        // one. It could be that this fails, and we notice that someone else
+                        // already minted a binding. In that case, update our view of the consensus
+                        // data and try again.
+
+                        let binding = 'outer: loop {
+                            let now = (self.now_fn)();
+                            let now_clamped =
+                                now - (now % self.timestamp_interval.as_millis() as u64);
+                            let mut current_ts = self.current_ts.lock().await;
+                            if now_clamped != *current_ts {
+                                *current_ts = now_clamped;
+                            }
+
+                            let new_binding = (*upper_element, *current_ts);
+                            let append_result = self
+                                .consensus
+                                .append(
+                                    vec![(new_binding, *current_version)],
+                                    Antichain::from_elem(*current_version),
+                                    Antichain::from_elem(*current_version + 1),
+                                )
+                                .await;
+
+                            let consensus_version = match append_result {
+                                Ok(_) => {
+                                    // We successfully advanced to a new version.
+                                    *current_version = *current_version + 1;
+                                    break new_binding;
```

Comment:
I didn't even know this syntax!

is it tricky to break out this loop into a method so these are just returns?

Reply:
Yeah, I'm not happy with this method, but didn't yet take any time to make it pretty.